util/app/discovery/discovery.go

```diff
@@ -62,6 +64,32 @@ func Discover(ctx context.Context, appPath, repoPath string, enableGenerateManif
 		if kustomize.IsKustomization(base) && IsManifestGenerationEnabled(v1alpha1.ApplicationSourceTypeKustomize, enableGenerateManifests) {
 			apps[dir] = string(v1alpha1.ApplicationSourceTypeKustomize)
 		}
+
+		if (strings.HasSuffix(base, ".yaml") || strings.HasSuffix(base, ".yml")) && IsManifestGenerationEnabled(v1alpha1.ApplicationSourceTypeDirectory, enableGenerateManifests) {
+			f, err := os.Open(path)
+			if err != nil {
+				log.Errorf("failed to open %s: %v", path, err)
+			}
+			defer f.Close()
+
+			decoder := yaml.NewDecoder(f)
+			obj := make(map[string]interface{})
+			err = decoder.Decode(&obj)
+			if err != nil {
+				log.Errorf("failed to decode %s: %v", path, err)
+			} else {
+				if obj["kind"] != nil && obj["apiVersion"] != nil && obj["metadata"] != nil {
+					// if plain manifest for type Directory is inside 'templates' dir in helm app, we should not consider it as type 'Directory'
+					// ex. if current dir is dir1/templates , dir1 will be checked first for helm app and later, dir1/templates must not be considered
+					parentDir, _ := filepath.Rel(appPath, filepath.Dir(filepath.Dir(path)))
```

Comment:
Is there an unhandled error here we should be handling?

Reply:
Have added logging for this.